src/ringmesh/geomodel/geomodel_builder.cpp

```diff
@@ -995,8 +1022,275 @@ namespace RINGMesh
+                    minimal_angle_line_id = line.index();
+                }
+            }
+            return minimal_angle_line_id;
+        }
+
+        OrientedLine get_next_surface_boundary_line(
+            const OrientedLine cur_line_and_side )
+        {
+            const auto& cur_line =
+                this->geomodel_.line( cur_line_and_side.index );
+            index_t cur_corner_id{
+                cur_line.boundary_gmme( cur_line_and_side.side ? 1 : 0 ).index()
+            };
+            const auto& cur_corner = this->geomodel_.corner( cur_corner_id );
+            if( cur_corner.nb_incident_entities() == 1 )
+            {
+                // Case the current line is an internal border of a
+                // surface
+                return OrientedLine(
+                    cur_line_and_side.index, !cur_line_and_side.side );
+            }
+
+            index_t next_line_id = find_line_with_minimal_angle(
+                cur_corner, cur_line_and_side.index );
+            ringmesh_assert( next_line_id != NO_ID );
+            bool next_side =
+                this->geomodel_.line( next_line_id ).boundary_gmme( 0 )
+                == cur_corner.gmme();
+            return OrientedLine( next_line_id, next_side );
+        }
+
+        std::vector< OrientedLine > get_surface_boundaries(
+            const index_t cur_surface_id,
+            std::vector< index_t >& line_indicent_surfaces )
+        {
+            std::vector< OrientedLine > cur_surface_boundaries;
+            OrientedLine first_line_and_side =
+                get_first_undetermined_line_side( line_indicent_surfaces );
+
+            OrientedLine cur_line_and_side{ first_line_and_side };
+            bool stop{ false };
+
+            // From the first line, the lines are walked turning around corners
+            // in the same direction (clockwise). Line after line, the
+            // boundaries
+            // of the currently processed surface are found until the algorithm
+            // has gone back to the first line.
+            // By convention, if a line is walked from its first boundary
+            // towards
+            // its second boundary, the surface is set as incident by the + side
+            // of the line.
+            while( !stop )
+            {
+                ringmesh_assert(
+                    line_indicent_surfaces[2 * cur_line_and_side.index
+                                           + ( cur_line_and_side.side ? 0 : 1 )]
+                    == NO_ID );
+                line_indicent_surfaces[2 * cur_line_and_side.index
+                                       + ( cur_line_and_side.side ? 0 : 1 )] =
+                    cur_surface_id;
+                cur_surface_boundaries.emplace_back( cur_line_and_side );
+
+                OrientedLine next_line_and_side =
+                    get_next_surface_boundary_line( cur_line_and_side );
+
+                if( next_line_and_side == first_line_and_side )
```

Comment:
Could you use this test directly into your loop? for example do { } while(  next_line_and_side != first_line_and_side );